File: packages-gen/dart/event/lib/util-decorate.dart
import 'package:xtalk_lang/common-data.dart' as xtd;

incr_fn() {
  var state = <dynamic, dynamic>{"value":-1};
  var next_id_fn = () {
    var i = 1 + state["value"];
    state["value"] = i;
    return "id-" + (i).toString();
  };
  return next_id_fn;
}

plugin_timing(handle) {
  var result = <dynamic, dynamic>{"output":<dynamic, dynamic>{}};
  var on_setup = (args) {
    return result["output"] = <dynamic, dynamic>{"start":DateTime.now().millisecondsSinceEpoch};
  };
  var on_teardown = () {
    var output = result["output"];
    var t_end = DateTime.now().millisecondsSinceEpoch;
    var t_elapsed = t_end - output["start"];
    output["end"] = t_end;
    return output["elapsed"] = t_elapsed;
  };
  var on_reset = () {
    return result["output"] = <dynamic, dynamic>{};
  };
  return xtd.obj_assign(result,<dynamic, dynamic>{
    "name":"timing",
    "on_setup":on_setup,
    "on_teardown":on_teardown,
    "on_reset":on_reset
  });
}

plugin_counts(handle) {
  var result = <dynamic, dynamic>{"output":<dynamic, dynamic>{"success":0,"error":0}};
  var on_success = (ret) {
    var output = result["output"];
    return output["success"] = (1 + output["success"]);
  };
  var on_error = (ret) {
    var output = result["output"];
    return output["error"] = (1 + output["error"]);
  };
  var on_reset = () {
    return result["output"] = <dynamic, dynamic>{"success":0,"error":0};
  };
  return xtd.obj_assign(result,<dynamic, dynamic>{
    "name":"counts",
    "on_success":on_success,
    "on_error":on_error,
    "on_reset":on_reset
  });
}

to_handle_callback(cb) {
  if(null == cb){
    cb = <dynamic, dynamic>{};
  }
  return <dynamic, dynamic>{
    "on_success":cb["success"],
    "on_error":cb["error"],
    "on_teardown":cb["finally"]
  };
}

new_handle(handler, plugin_fns, opts) {
  var create_fn = opts["create_fn"];
  var delay = opts["delay"];
  var dump_fn = opts["dump_fn"];
  var id_fn = opts["id_fn"];
  var name = opts["name"];
  var wrap_fn = opts["wrap_fn"];
  if(null == id_fn){
    id_fn = Function.apply((incr_fn as Function),<dynamic>[]);
  }
  if(null == create_fn){
    create_fn = ((x) {
      return x;
    });
  }
  if(null == delay){
    delay = 0;
  }
  var handle = Function.apply((create_fn as Function),<dynamic>[
    <dynamic, dynamic>{
      "::":"handle",
      "name":name,
      "id_fn":id_fn,
      "wrap_fn":wrap_fn,
      "handler":handler,
      "delay":delay
    }
  ]);
  var plugins = xtd.arr_map(plugin_fns,(f) {
    return f(handle);
  });
  handle["plugins"] = plugins;
  return handle;
}

run_handle(handle, args, tcb) {
  var delay = handle["delay"];
  var handler = handle["handler"];
  var id_fn = handle["id_fn"];
  var plugins = handle["plugins"];
  var wrap_fn = handle["wrap_fn"];
  if(null == args){
    args = <dynamic>[];
  }
  var tcbs = xtd.arr_clone(plugins);
  if(null != tcb){
    if((tcb.runtimeType).toString().startsWith("List") || (tcb.runtimeType).toString().startsWith("_GrowableList")){
      xtd.arr_assign(tcbs,tcb);
    }
    else{
      tcbs.add(tcb);
    }
  }
  if((delay.runtimeType).toString().contains("Function") || (delay.runtimeType).toString().contains("=>") || (delay).toString().startsWith("Closure")){
    delay = delay();
  }
  var receipt = <dynamic, dynamic>{"id":Function.apply((id_fn as Function),<dynamic>[])};
  var teardown_fn = () {
    var arr_41388 = tcbs;
    for(var i41389 = 0; i41389 < arr_41388.length; ++i41389){
      var cb = arr_41388[i41389];
      var on_teardown = cb["on_teardown"];
      var name = cb["name"];
      var output = cb["output"];
      if(null != on_teardown){
        on_teardown();
      }
      if((null != name) && (null != output)){
        receipt[name] = output;
      }
    };
    return receipt;
  };
  var run_fn = () {
    var arr_41410 = tcbs;
    for(var i41411 = 0; i41411 < arr_41410.length; ++i41411){
      var cb = arr_41410[i41411];
      var on_setup = cb["on_setup"];
      if(null != on_setup){
        on_setup(args);
      }
    };
    var base_thunk = () {
      return Function.apply(handler,args);
    };
    var base_promise = (0 < ((null == delay) ? 0 : delay)) ? Future.delayed(Duration(milliseconds:  delay )).then((_) {
      return Future.sync(() {
        return Function.apply(base_thunk,<dynamic>[]);
      });
    }) : Future.sync(base_thunk);
    return ((Future.sync(() => (() async { try { return await ((Future.sync(() => ((Future.sync(() => base_promise)) as Future<dynamic>).then((value) async { return await Function.apply((ret) {
      var arr_41440 = tcbs;
      for(var i41441 = 0; i41441 < arr_41440.length; ++i41441){
        var cb = arr_41440[i41441];
        var on_success = cb["on_success"];
        if(null != on_success){
          on_success(ret);
        }
      };
      receipt["status"] = "success";
      receipt["value"] = ret;
      return receipt;
    },<dynamic>[value]); }))) as Future<dynamic>); } catch (err) { return await Future.sync(() => Function.apply((err) {
      var arr_41462 = tcbs;
      for(var i41463 = 0; i41463 < arr_41462.length; ++i41463){
        var cb = arr_41462[i41463];
        var on_error = cb["on_error"];
        if(null != on_error){
          on_error(err);
        }
      };
      receipt["status"] = "error";
      receipt["error"] = err;
      var reject_fn = () {
        throw receipt;
      };
      return Future.sync(reject_fn);
    },<dynamic>[err])); } })())) as Future<dynamic>).whenComplete(() async { await Function.apply(teardown_fn,<dynamic>[]); });
  };
  var proc = (null != wrap_fn) ? Function.apply((wrap_fn as Function),<dynamic>[run_fn,args,receipt,handle]) : Function.apply((run_fn as Function),<dynamic>[]);
  return Future.sync(() {
    return proc;
  });
}

File: packages-gen/js/event/util-decorate.js
const xtd = require("@xtalk/lang/common-data.js")

function incr_fn(){
  let state = {"value":-1};
  let next_id_fn = function (){
    let i = 1 + state["value"];
    state["value"] = i;
    return "id-" + String(i);
  };
  return next_id_fn;
}

function plugin_timing(handle){
  let result = {"output":{}};
  let on_setup = function (args){
    result["output"] = {"start":Date.now()};
  };
  let on_teardown = function (){
    let output = result["output"];
    let t_end = Date.now();
    let t_elapsed = t_end - output["start"];
    output["end"] = t_end;
    output["elapsed"] = t_elapsed;
  };
  let on_reset = function (){
    result["output"] = {};
  };
  return Object.assign(result,{
    "name":"timing",
    "on_setup":on_setup,
    "on_teardown":on_teardown,
    "on_reset":on_reset
  });
}

function plugin_counts(handle){
  let result = {"output":{"success":0,"error":0}};
  let on_success = function (ret){
    let {output} = result;
    output["success"] = (1 + output["success"]);
  };
  let on_error = function (ret){
    let {output} = result;
    output["error"] = (1 + output["error"]);
  };
  let on_reset = function (){
    result["output"] = {"success":0,"error":0};
  };
  return Object.assign(result,{
    "name":"counts",
    "on_success":on_success,
    "on_error":on_error,
    "on_reset":on_reset
  });
}

function to_handle_callback(cb){
  if(null == cb){
    cb = {};
  }
  return {
    "on_success":cb["success"],
    "on_error":cb["error"],
    "on_teardown":cb["finally"]
  };
}

function new_handle(handler,plugin_fns,opts){
  let {create_fn,delay,dump_fn,id_fn,name,wrap_fn} = opts;
  if(null == id_fn){
    id_fn = incr_fn();
  }
  if(null == create_fn){
    create_fn = (function (x){
      return x;
    });
  }
  if(null == delay){
    delay = 0;
  }
  let handle = create_fn({
    "::":"handle",
    "name":name,
    "id_fn":id_fn,
    "wrap_fn":wrap_fn,
    "handler":handler,
    "delay":delay
  });
  let plugins = plugin_fns.map(function (f){
    return f(handle);
  });
  handle["plugins"] = plugins;
  return handle;
}

function run_handle(handle,args,tcb){
  let {delay,handler,id_fn,plugins,wrap_fn} = handle;
  if(null == args){
    args = [];
  }
  let tcbs = plugins.slice();
  if(null != tcb){
    if(Array.isArray(tcb)){
      xtd.arr_assign(tcbs,tcb);
    }
    else{
      tcbs.push(tcb);
    }
  }
  if("function" == (typeof delay)){
    delay = delay();
  }
  let receipt = {"id":id_fn()};
  let teardown_fn = function (){
    for(let cb of tcbs){
      let on_teardown = cb["on_teardown"];
      let name = cb["name"];
      let output = cb["output"];
      if(null != on_teardown){
        on_teardown();
      }
      if((null != name) && (null != output)){
        receipt[name] = output;
      }
    };
    return receipt;
  };
  let run_fn = function (){
    for(let cb of tcbs){
      let on_setup = cb["on_setup"];
      if(null != on_setup){
        on_setup(args);
      }
    };
    let base_thunk = function (){
      return handler.apply(null,args);
    };
    let base_promise = (0 < ((null == delay) ? 0 : delay)) ? new Promise(function (resolve,reject){
      setTimeout(function (){
        new Promise(function (inner_resolve){
          inner_resolve(base_thunk());
        }).then(function (value){
          resolve(value);
        }).catch(function (err){
          reject(err);
        });
      },delay);
    }) : Promise.resolve().then(base_thunk);
    return base_promise.then(function (ret){
      for(let cb of tcbs){
        let on_success = cb["on_success"];
        if(null != on_success){
          on_success(ret);
        }
      };
      receipt["status"] = "success";
      receipt["value"] = ret;
      return receipt;
    }).catch(function (err){
      for(let cb of tcbs){
        let on_error = cb["on_error"];
        if(null != on_error){
          on_error(err);
        }
      };
      receipt["status"] = "error";
      receipt["error"] = err;
      let reject_fn = function (){
        throw receipt;
      };
      return Promise.resolve().then(reject_fn);
    }).finally(teardown_fn);
  };
  let proc = (null != wrap_fn) ? wrap_fn(run_fn,args,receipt,handle) : run_fn();
  return Promise.resolve().then(function (){
    return proc;
  });
}

module.exports = {
  ["incr_fn"]:incr_fn,
  ["plugin_timing"]:plugin_timing,
  ["plugin_counts"]:plugin_counts,
  ["to_handle_callback"]:to_handle_callback,
  ["new_handle"]:new_handle,
  ["run_handle"]:run_handle
}